### DISCUSSION TITLE
DataGrid inside Expander or StackPanel gets poor/inconsistent sizing

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#10182] [Q&A]
A DataGrid with row details doesn't size within its containing layout controls properly. It creates its own scroll bars although there is plenty of vertical space. Are these bugs, or is there something in the XAML I can do to force the DataGrid to fully grow (leaving scrolling to its container) rather than create a vertical scroll bar?

Repro code is below, and screenshots:

Initially, the data rows aren't visible at all:
![image](https://user-images.githubusercontent.com/3548/216501340-1a2bbd29-2462-45cc-8177-e412bb10fa90.png)

Clicking the header gets the data rows to show up:
![image](https://user-images.githubusercontent.com/3548/216501392-5cc37a10-ea8d-4a22-9b24-d1e4085785bf.png)

Once clicking on any of the data rows, the details expands and scrollbars needlessly appear. 
![image](https://user-images.githubusercontent.com/3548/216501415-d2256087-c9c8-4074-86dc-2190e1ee3cb5.png)

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:AvaloniaApplication1.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="300" d:DesignHeight="450"
		Width="300" Height="450"
        x:Class="AvaloniaApplication1.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="AvaloniaApplication1">

	<Window.DataContext>
		<vm:MainWindowViewModel/>
	</Window.DataContext>

	<StackPanel>
		<Expander Header="Expander" IsExpanded="True">
			<DataGrid Items="{Binding Rows}" AutoGenerateColumns="False" RowDetailsVisibilityMode="VisibleWhenSelected" SelectionMode="Extended">
				<DataGrid.RowDetailsTemplate>
					<DataTemplate>
						<StackPanel>
							<TextBlock Text="Line 1" />
							<TextBlock Text="Line 2" />
						</StackPanel>
					</DataTemplate>
				</DataGrid.RowDetailsTemplate>
				<DataGrid.Columns>
					<DataGridTextColumn Binding="{Binding Name}" Header="Name" Width="*" />
				</DataGrid.Columns>
			</DataGrid>
		</Expander>
	</StackPanel>
</Window>
```

View model:
```cs
using DynamicData;
using ReactiveUI;
using System.Collections.ObjectModel;
using System.Linq;

namespace AvaloniaApplication1.ViewModels;

public class MainWindowViewModel : ViewModelBase {
    public MainWindowViewModel()
    {
        this.Rows.AddRange(Enumerable.Range(1, 5).Select(i => new Person { Name = $"Name {i}" }));
    }

    public ObservableCollection<Person> Rows { get; } = new();
}

public class Person : ViewModelBase {
    private string name = string.Empty;
    
    private bool b;

    public string Name {
        get => this.name;
        set => this.RaiseAndSetIfChanged(ref this.name, value);
    }

    public bool Checked {
        get => this.b;
        set => this.RaiseAndSetIfChanged(ref this.b, value);
    }
}
```

Note that multi-select gets some interesting effects. Also, removing the StackPanel and/or Expander around the DataGrid manifest other weird layout/position behaviors.

**Comment by timunie** [comment]
StackPanel has infite height, so DataGrid will size into non visible area. You can either limit the heigt of the DataGrid or place it in a different Panel like a Grid or DockPanel.

**Reply by AArnott** [reply]
> so DataGrid will size into non visible area

Of itself? The problem is that the DataGrid seems to be limiting its own height, adding its own scrollbar since it has more content than will fit into its own height constraint. If I set the Height on the DataGrid itself to force it to be larger, the scrollbar vanishes and things look better. But of course in general I don't know what height I should set because I can't predict how many rows will appear. And with row details that change in visibility, the height can change over time.

**Reply by timunie** [reply]
DataGrid has it's own ScrollViewer internally. You may not place it in an outer scrollviewer or Stackpanel unlesss you limit it's height. MaxHeight should also work.

**Reply by AArnott** [reply]
Setting `MaxHeight` didn't solve the problem. It didn't have any effect that I could detect. Setting `Height` explicitly does solve it, except that now the outer ScrollViewer believes it needs to offer far more scrolling into emptiness than necessary.

I got rid of the StackPanel and ScrollViewer, but kept the Expander because that's fundamental to the UX in my window. The DataGrid still doesn't grow to fill its container. Sticking it inside a Grid didn't help either.

**Reply by timunie** [reply]
It works for me if I define `VerticalAlignment` of the `Exapnder` and Remove the outer StackPanel:

```diff
-   <StackPanel>
        <Expander Header="Expander" IsExpanded="True"
+                  VerticalAlignment="Stretch">
            <DataGrid 
                Items="{Binding Rows}" AutoGenerateColumns="False" RowDetailsVisibilityMode="VisibleWhenSelected" SelectionMode="Extended">
                <DataGrid.RowDetailsTemplate>
                    <DataTemplate>
                        <StackPanel>
                            <TextBlock Text="Line 1" />
                            <TextBlock Text="Line 2" />
                        </StackPanel>
                    </DataTemplate>
                </DataGrid.RowDetailsTemplate>
                <DataGrid.Columns>
                    <DataGridTextColumn Binding="{Binding Name}" Header="Name" Width="*" />
                </DataGrid.Columns>
            </DataGrid>
        </Expander>
-   </StackPanel>
```

I don't know if that is what you are looking for. maybe you need to change the Style of the expander to only Stratch if expanded:

```xml
<Window.Styles>
    <Style Selector="Expander">
        <Setter Property="VerticalAlignment" Value="Top" />
    </Style>
    <Style Selector="Expander:expanded">
        <Setter Property="VerticalAlignment" Value="Stretch" />
    </Style>
</Window.Styles>
```

Happy coding
Tim

**Reply by AArnott** [reply]
Thanks for the suggestions. I tried this but it ultimately didn't work. I have two expanders, one above the other, so I need some kind of flexible layout control around both of them. I tried a grid again with rows either sized or auto height and the effect wasn't good.
For now, I've switched to the TabControl for the two regions. The UX isn't as good as two expanders would have been, but at least scrolling is avoided.

**Reply by timunie** [reply]
```xml
    <Style Selector="Expander:expanded">
        <Setter Property="VerticalAlignment" Value="Top" />
        <Setter Property="MinHeight" Value="300" />
        <Setter Property="MaxHeight" Value="600" />
    </Style>
```
just another idea
